### PR TITLE
Adding inverse kinematic controller 

### PR DESCRIPTION
**Pull request by CarlottaSartore**
This PR include the inverse kinematic controller and the related example to make the robot walk, in position control, using  bipedal locomotion inverse kinematic, centroidal MPC  and mujoco simulator. 
It depends on bipedal locomotion framework version 0.16

### CODE REVIEW

**Comment by CarlottaSartore** [comment]
This PR is still a draft, because we need to understand how to automatically add tag to the contacts in mujoco for the feet rear and front.

**Comment by CarlottaSartore** [comment]
> This PR is still a draft, because we need to understand how to automatically add tag to the contacts in mujoco for the feet rear and front.

This has been solved in 25ada23898c19d449d2d74c7bc5674b2c289dda2 , but there are still some issues in the computation of the wrenches that I want to furtherly investigate  before resolving the draft status

**Comment by CarlottaSartore** [comment]
I have decided to add in a future a PR solely for the examples. thus now this PR is ready for review. 

The problem with the wrenches has been solved. 
In addition we have added: 
- Automatic load of meshes 
- Automatic mapping in between input `joint name list` and mujoco order

File: README.md [hunk 1]
```diff
@@ -32,7 +32,7 @@ To install you can use the following commands
 
 
 ```
-conda create -n comododev -c conda-forge adam-robotics idyntree bipedal-locomotion-framework=0.15.0 mujoco mujoco-python numpy mujoco-python-viewer matplotlib urllib3 urchin
+conda create -n comododev -c conda-forge adam-robotics idyntree bipedal-locomotion-framework=0.16.0 mujoco mujoco-python numpy mujoco-python-viewer matplotlib urllib3 urchin resolve-robotics-uri-py
```
Comment (traversaro) [inline README.md:35]: Do we need to hardcode blf version? Can we just specify a minimum version?

File: src/comodo/robotModel/robotModel.py [hunk 2]
```diff
@@ -188,43 +197,78 @@ def rotation_matrix_to_quaternion(self, R):
...
-        ## Adding compiler discard visaul false for mujoco rendering
         mujoco_el = ET.Element("mujoco")
         compiler_el = ET.Element("compiler")
         compiler_el.set("discardvisual", "false")
+        if not (meshes is None):
+            compiler_el.set("meshdir", str(meshes))
         mujoco_el.append(compiler_el)
         robot_el.append(mujoco_el)
-        # Convert the XML tree to a string
+        # # Convert the XML tree to a string
```
Comment (traversaro) [inline src/comodo/robotModel/robotModel.py:264]: Is this change intended?

**Review by traversaro — APPROVED** [review]
Minor comments.

**Comment by S-Dafarra** [comment]
Nit. ``kinematic`` is and adjective, while ``kinematics`` is a noun. In this case I think it should be ``InverseKinematics``

File: examples/mujoco_walking_position.ipynb [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    "while t < TIME_TH:\n",
+    "    # Reading robot state from simulator\n",
+    "    s, ds, tau = mujoco_instance.get_state()\n",
+    "    energy_i = np.linalg.norm(tau)\n",
+    "    H_b = mujoco_instance.get_base()\n",
+    "    w_b = mujoco_instance.get_base_velocity()\n",
+    "    t = mujoco_instance.get_simulation_time()\n",
+    "    if H_b[2, 3] < 0.5:\n",
+    "        robot_standing = False\n",
+    "        print(\"Robot felt\")\n",
```
Comment (S-Dafarra) [inline examples/mujoco_walking_position.ipynb:235]:
```suggestion
    "        print(\"Robot fell\")\n",
```

**Review by S-Dafarra — COMMENTED** [review]
Some initial comments on the notebook

File: src/comodo/mujocoSimulator/mujocoSimulator.py [hunk 4]
```diff
@@ -26,6 +32,16 @@ def load_model(self, robot_model, s, xyz_rpy, kv_motors=None, Im=None):
...
         )
+        self.H_left_foot = copy.deepcopy(self.robot_model.H_left_foot)
+        self.H_right_foot = copy.deepcopy(self.robot_model.H_right_foot)
+        self.H_left_foot_num = None 
+        self.H_right_foot_num = None 
+
+    def get_contact_status(self):
+        left_wrench, rigth_wrench = self.get_feet_wrench()
+        left_foot_contact = left_wrench[2] > 5
+        right_foot_contact = rigth_wrench[2] > 5
```
Comment (S-Dafarra) [inline src/comodo/mujocoSimulator/mujocoSimulator.py:43]: 5 is a magic number? 😊

File: src/comodo/mujocoSimulator/mujocoSimulator.py [hunk 5]
```diff
@@ -34,41 +50,90 @@ def set_visualize_robot_flag(self, visualize_robot):
 
     def set_base_pose_in_mujoco(self, xyz_rpy):
         base_xyz_quat = np.zeros(7)
-        base_xyz_quat[:3] = xyz_rpy[:3]
+        base_xyz_quat[:3] = xyz_rpy[:3] + 0.0001
```
Comment (S-Dafarra) [inline src/comodo/mujocoSimulator/mujocoSimulator.py:53]: This looks weird. Are you setting a quaternion from an rpy? Also what is the 0.0001 for?

File: src/comodo/robotModel/robotModel.py [hunk 6]
```diff
@@ -22,14 +23,28 @@ def __init__(
...
         right_foot: str = "r_sole",
         torso: str = "chest",
+        right_foot_rear_link_name: str = "r_foot_rear",
+        right_foot_front_link_name: str = "r_foot_front",
+        left_foot_rear_link_name: str = "l_foot_rear",
+        left_foot_front_link_name: str = "l_foot_front",
+        kp_pos_control: np.float32 = np.array(
+            [35 * 70.0, 35 * 70.0, 35 * 40.0, 35 * 100.0, 35 * 100.0, 35 * 100.0,35 * 70.0, 35 * 70.0, 35 * 40.0, 35 * 100.0, 35 * 100.0, 35 * 100.0,20 * 5.745, 20 * 5.745, 20 * 5.745, 20 * 1.745,20 * 5.745, 20 * 5.745, 20 * 5.745, 20 * 1.745]
+        ),
+        kd_pos_control: np.float32= np.array([15 * 0.15, 15 * 0.15, 15 * 0.35, 15 * 0.15, 15 * 0.15, 15 * 0.15,15 * 0.15, 15 * 0.15, 15 * 0.35, 15 * 0.15, 15 * 0.15, 15 * 0.15,4 * 5.745, 4 * 5.745, 4 * 5.745, 4 * 1.745,4 * 5.745, 4 * 5.745, 4 * 5.745, 4 * 1.745])
```
Comment (S-Dafarra) [inline src/comodo/robotModel/robotModel.py:33]: To be honest, this seems a bit ergoCub specific. Is this wanted?

Comment (S-Dafarra) [inline src/comodo/robotModel/robotModel.py:33]: Why not using directly the resulting numbers instead of leaving the multiplications?

File: examples/mujoco_walking_position.ipynb [hunk 7]
```diff
@@ -0,0 +1,325 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## (W.I.P.)Mujoco with IK and MPC Example \n",
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply examples/mujoco_walking_position.ipynb:7]: The example is a work in progress, I have decided to incorporate all the examples in a unique PR that I will open after this. 
The example is working, but the gain must be identified to make the robot walk, this is also specified in the example explanation

File: examples/mujoco_walking_position.ipynb [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    "This examples, load a basic robot model (i.e. composed only of basic shapes), modifies the links of such a robot model by elongating the legs, define instances of the IK (Inverse Kinematic) and Centroidal MPC  controller and simulate the behavior of the robot using mujoco.  \n",
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply examples/mujoco_walking_position.ipynb]: Thanks :)

File: examples/mujoco_walking_position.ipynb [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    "\n",
+    ":warning: this is still a work in progress. Missing tuning on the controller to perfrom few steps "
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply examples/mujoco_walking_position.ipynb]: Thanks :)

File: src/comodo/mujocoSimulator/mujocoSimulator.py (hunk 5 above)
Comment (CarlottaSartore) [reply src/comodo/mujocoSimulator/mujocoSimulator.py:53]: `base_xyz_quat` contains the base XYZ position and the orientation as  quaternion, in the order xyz_quat
meanwhile `xyz_rpy` contains the base `xyz` position and orientation as `rpy` in the order xyz_rpy. 
Therefore I am setting the XYZ position of the base, adding  `0.0001` to the z component to lift from the ground (even if the value is un-appreciable )

File: examples/mujoco_walking_position.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (S-Dafarra) [reply examples/mujoco_walking_position.ipynb:7]: I see, just wanted to make sure it was not a leftover

File: src/comodo/robotModel/robotModel.py (hunk 6 above)
Comment (CarlottaSartore) [reply src/comodo/robotModel/robotModel.py:33]: This is a leftover of the tuning process. I prefer to leave like this since the low-level position control is not really stable

File: src/comodo/mujocoSimulator/mujocoSimulator.py (hunk 5 above)
Comment (S-Dafarra) [reply src/comodo/mujocoSimulator/mujocoSimulator.py:53]: Ah I understand. Still the 0.0001 is a bit of a magic number. If it is needed to make things working, then I would suggest to promote it as a parameter.

File: src/comodo/robotModel/robotModel.py (hunk 6 above)
Comment (S-Dafarra) [reply src/comodo/robotModel/robotModel.py:33]: What happens if the foot is not split in two parts?

File: src/comodo/mujocoSimulator/mujocoSimulator.py (hunk 4 above)
Comment (CarlottaSartore) [reply src/comodo/mujocoSimulator/mujocoSimulator.py]: Improved in 6090ad00c61c867ae6d8e7e0954bdcb4aa298d6a (dependend on the robot mass )

File: src/comodo/inverseKinematic/inverseKinematicParamTuning.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+class InverseKinematicParamTuning:
+    def __init__(self) -> None:
+        self.zmp_gain = [0.4, 0.4]
+        self.com_gain = [4.3, 4.3]
+        self.foot_linear = 1.0
+        self.foot_angular = 6.59
+        self.com_linear = 2.59
+        self.chest_angular = 1.4
+        self.root_linear = 2.8
+        self.Kp_joint_tracking = 5.0
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply src/comodo/inverseKinematic/inverseKinematicParamTuning.py]: Done in 6090ad00c61c867ae6d8e7e0954bdcb4aa298d6a

File: examples/mujoco_walking_position.ipynb [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    "# Load the URDF file\n",
+    "tree = ET.parse(urdf_robot_file.name)\n",
+    "root = tree.getroot()\n",
+    "\n",
+    "# Convert the XML tree to a string\n",
+    "robot_urdf_string_original = ET.tostring(root)\n",
+    "\n",
+    "create_urdf_instance = createUrdf(\n",
+    "    original_urdf_path=urdf_robot_file.name, save_gazebo_plugin=False\n",
+    ")"
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply examples/mujoco_walking_position.ipynb:63]: I prefer to avoid it. I would like to encourage the usage of comodo with string objects rather than files.  I prefer not to give too many utilities for converting string model to file and from file to encourage to avoid an excessive use of such functions that could create big issues (passing from file several time )

File: README.md (hunk 1 above)
Comment (CarlottaSartore) [reply README.md:35]: I prefer to be dependent on a specific version since blf is continuously changing and I do not have any CI yet to check that the code works properly

File: src/comodo/inverseKinematic/inverseKinematic.py [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+        Parameters:
+        - position (list or numpy array): 3D position [x, y, z]
+        - quaternion (list or numpy array): Quaternion [w, x, y, z]
+
+        Returns:
+        - transformation_matrix (numpy array): 4x4 transformation matrix
+        """
+
+        # Normalize quaternion to ensure it's a unit quaternion
+        # quaternion = quaternion / np.linalg.norm(quaternion)
```
[the remark this answers is not in the document]
Comment (CarlottaSartore) [reply src/comodo/inverseKinematic/inverseKinematic.py]: Done in 5e1463e5fe2e055123323ca32feafde5da7bda03

File: src/comodo/robotModel/robotModel.py (hunk 6 above)
Comment (CarlottaSartore) [reply src/comodo/robotModel/robotModel.py:33]: I agree this could be improved (By passing a list of link names per foot). Anyhow, I have currently a few already implemented PR  to open and this change would be breaking. I prefer to implement this enhancement when I finalize the other review. I have opened an issue to track this: https://github.com/ami-iit/comodo/issues/14

**Comment by CarlottaSartore** [comment]
@S-Dafarra  I think I resolve all your comments, let me know if there is something I have missed

**Review by S-Dafarra — APPROVED** [review]

**Comment by CarlottaSartore** [comment]
Thanks all ! Merging !
